File: wpilibjExamples/src/test/java/edu/wpi/first/wpilibj/examples/addressableled/RainbowTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.wpilibj.examples.addressableled;

import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import edu.wpi.first.hal.HAL;
import edu.wpi.first.wpilibj.simulation.AddressableLEDSim;
import edu.wpi.first.wpilibj.util.Color;
import edu.wpi.first.wpilibj.util.Color8Bit;
import org.junit.jupiter.api.Test;

class RainbowTest {
  @Test
  void rainbowPatternTest() {
    HAL.initialize(500, 0);
    try (var robot = new Robot()) {
      robot.robotInit();
      AddressableLEDSim ledSim = AddressableLEDSim.createForChannel(9);
      assertTrue(ledSim.getRunning());
      assertEquals(60, ledSim.getLength());

      var rainbowFirstPixelHue = 0;
      for (int iteration = 0; iteration < 100; iteration++) {
        robot.robotPeriodic();
        var data = ledSim.getData();
        for (int i = 0; i < 60; i++) {
          final var hue = (rainbowFirstPixelHue + (i * 180 / 60)) % 180;
          assertIndexColor(data, i, hue, 255, 128);
        }
        rainbowFirstPixelHue += 3;
        rainbowFirstPixelHue %= 180;
      }
    }
  }

  private void assertIndexColor(byte[] data, int index, int hue, int saturation, int value) {
    var color = new Color8Bit(Color.fromHSV(hue, saturation, value));
    int b = data[index * 4];
    int g = data[(index * 4) + 1];
    int r = data[(index * 4) + 2];
    int z = data[(index * 4) + 3];

    assertAll(
        () -> assertEquals(0, z),
        () -> assertEquals(color.red, r & 0xFF),
        () -> assertEquals(color.green, g & 0xFF),
        () -> assertEquals(color.blue, b & 0xFF));
  }
}

File: wpilibcExamples/src/test/cpp/examples/AddressableLED/cpp/RainbowTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <array>

#include <fmt/format.h>
#include <frc/AddressableLED.h>
#include <frc/simulation/AddressableLEDSim.h>
#include <frc/util/Color.h>
#include <frc/util/Color8Bit.h>
#include <gtest/gtest.h>
#include <hal/AddressableLEDTypes.h>

#include "Robot.h"

void AssertIndexColor(std::array<frc::AddressableLED::LEDData, 60> data,
                      int index, int hue, int saturation, int value);

TEST(RainbowTest, RainbowPattern) {
  Robot robot;
  robot.RobotInit();
  frc::sim::AddressableLEDSim ledSim =
      frc::sim::AddressableLEDSim::CreateForChannel(9);
  EXPECT_TRUE(ledSim.GetRunning());
  EXPECT_EQ(60, ledSim.GetLength());

  auto rainbowFirstPixelHue = 0;
  std::array<frc::AddressableLED::LEDData, 60> data;
  for (int iteration = 0; iteration < 100; iteration++) {
    SCOPED_TRACE(fmt::format("Iteration {} of 100", iteration));
    robot.RobotPeriodic();
    ledSim.GetData(data.data());
    for (int i = 0; i < 60; i++) {
      SCOPED_TRACE(fmt::format("LED {} of 60", i));
      auto hue = (rainbowFirstPixelHue + (i * 180 / 60)) % 180;
      AssertIndexColor(data, i, hue, 255, 128);
    }
    rainbowFirstPixelHue += 3;
    rainbowFirstPixelHue %= 180;
  }
}

void AssertIndexColor(std::array<frc::AddressableLED::LEDData, 60> data,
                      int index, int hue, int saturation, int value) {
  frc::Color8Bit color{frc::Color::FromHSV(hue, saturation, value)};

  EXPECT_EQ(0, data[index].padding);
  EXPECT_EQ(color.red, data[index].r & 0xFF);
  EXPECT_EQ(color.green, data[index].g & 0xFF);
  EXPECT_EQ(color.blue, data[index].b & 0xFF);
}
